### DISCUSSION TITLE
Help with a layer

### DISCUSSION THREAD

**Discussion by emadb** [evan-liu/karabiner.ts#112] [Q&A] [answered]
I created this layer to use `i`,`j`,`k`,`l` as arrows when `caps_lock` is held:

```
layer('caps_lock', 'mod-layer')
  .manipulators([
    map('i').to('up_arrow'), 
    map('j').to('left_arrow'), 
    map('k').to('down_arrow'), 
    map('l').to('right_arrow'),
  ]),
```

It works perfectly.

Now I would like that pressing `a` while `caps_lock` is held the arrows can select text (`a` should map `left_shift`) to mimic the behaviour of SHIFT+Arrows to select.

I tried different ways (condition, rules, etc...) but I can't find the correct configuration.

**Comment by evan-liu** [comment]
There are a few options: 

- [duoLayer()](https://karabiner.ts.keymaps.app/rules/duo-layer) (one of the keys can be released once the layer is triggered)
- [modifierLayer()](https://karabiner.ts.keymaps.app/rules/hyper-layer)
- [pass modifiers to layer](https://github.com/evan-liu/karabiner.ts/issues/89#issuecomment-1597742983)

Please let me know if any of them help.

**Reply by emadb** [reply]
I try without succeed, but probably I didn't explain with enough details.
My actual config is this:

```typescript
simlayer('spacebar', 'mod-layer', 500)
  .options({
    key_down_order: 'strict',
  })
  .manipulators([
    map('a').to('left_shift'),               
    map('s').to('left_control'), 
    map('d').to('left_option'),  
    map('f').to('left_command'),            
  ]),

  layer('caps_lock', 'nav-layer')
  .configKey((v) => v.toIfAlone('escape'), true)
  .manipulators([
    map('i').to('up_arrow'), 
    map('j').to('left_arrow'), 
    map('k').to('down_arrow'), 
    map('l').to('right_arrow')
  ]),
```
I have two layers: one the modifiers (`ASDF`) and another for the arrows (`IJKL`)

Sometimes is useful to use SHIFT-Arrows to select text but my two layers are mutually exclusive. I would like to use `space`+`A`+`L` to select text on the right (emulate SHIFT-Right Arrow).

I know that with plain karabiner (JSON file) is possible but I cant' replicate with the DSL of Karabiner.ts

thank you

**Comment by evan-liu** [comment]
If you want to press space+A at the same time to trigger a layer, on which `i,j,k,l` map to selecting, you can use duoLayer: 

```typescript
  duoLayer('␣', 'a').manipulators({
    h: toKey('←', '⇧'),
    j: toKey('↓', '⇧'),
    k: toKey('↑', '⇧'),
    l: toKey('→', '⇧'),
  }),
```

If you want to press space first to trigger `mod-layer`, while holding the space, use `a` as shift when using `i,j,k,l`, you can try:

```typescript
  simlayer('spacebar', 'mod-layer', 500)
    .options({
      key_down_order: 'strict',
    })
    .manipulators([
      map('a').to('left_shift'),
      map('s').to('left_control'),
      map('d').to('left_option'),
      map('f').to('left_command'),

      map('h', '⇧').to('←', '⇧'),
      map('j', '⇧').to('↓', '⇧'),
      map('k', '⇧').to('↑', '⇧'),
      map('l', '⇧').to('→', '⇧'),
    ]),
```

If you want to press caps lock to trigger `nav-layer` first, instead of space / `mod-layer`, you can try: 

```typescript
  // **Note** Needs to be before
  // layer('caps_lock', 'nav-layer')
  layer('a', 'nav-select-layer')
    .condition(ifVar('nav-layer')) // `mod-layer` could also work
    .manipulators({
      h: toKey('←', '⇧'),
      j: toKey('↓', '⇧'),
      k: toKey('↑', '⇧'),
      l: toKey('→', '⇧'),
    }),
   
  // **Note** `nav-layer` needs to go after `nav-select-layer`
```

Let me know if it helps. If not, would it be possible for you to create a plain Karabiner JSON that explains what you need?

**Comment by emadb** [comment]
The first option with `duoLayer` works, the config becomes a little bit repetitive but I think I could refactor it to keep it readable and compact.

The vanilla Karabiner configuration that used to work is this:

```JSON
"rules": [
{
    "description": "CAPS LOCK + l to arrow keys",
    "manipulators": [
        {
            "conditions": [
                {
                    "name": "caps_lock pressed",
                    "type": "variable_if",
                    "value": 1
                }
            ],
            "from": {
                "key_code": "l",
                "modifiers": { "optional": [ "any" ]
                }
            },
            "to": [ { "key_code": "right_arrow" }
            ],
            "type": "basic"
        },
        {
            "from": {
                "key_code": "caps_lock",
                "modifiers": { "optional": [ "any" ] }
            },
            "to": [
                {
                    "set_variable": {
                        "name": "caps_lock pressed",
                        "value": 1
                    }
                }
            ],
            "to_after_key_up": [
                {
                    "set_variable": {
                        "name": "caps_lock pressed",
                        "value": 0
                    }
                }
            ],
            "type": "basic"
        }
    ]
},
{
    "description": "CAPS LOCK + a Shift",
    "manipulators": [
        {
            "conditions": [
                {
                    "name": "caps_lock pressed",
                    "type": "variable_if",
                    "value": 1
                }
            ],
            "from": {
                "key_code": "a",
                "modifiers": { "optional": [ "any" ] }
            },
            "to": [{ "key_code": "left_shift" }
            ],
            "type": "basic"
        }
    ]
}
]
```

I simplify it (it is just caps lock, A and L) so it is readable.
They are just two manipulators that works in parallel. 

Is there a way, with karabiner.ts to cascade the layers and let a key pass into different layers?

**Comment by evan-liu** [comment, marked as answer]
This produces similar JSON (try it at [playground](https://stackblitz.com/edit/karabiner-ts-fxi9s4?file=rules.js)): 

```typescript
  layer('⇪', 'caps_lock pressed')
    .modifiers('??')
    .manipulators({
      l: toKey('→'),
      a: toKey('‹⇧'),
    }),
```

Does it work?

**Reply by evan-liu** [reply]
I think the key part is `layer().modifiers('??')`, which makes the modifiers (`⇧` in this case, mapped from `a`) pass through.

**Reply by evan-liu** [reply]
It is a simple and elegant way to select, I've [changed my own config](https://github.com/evan-liu/karabiner-config/commit/d655e0bda9482405618a2ec95df137ad11a00116). Thanks 😄

**Reply by emadb** [reply]
Thank you, it works perfectly. I didn't know about the special chars `??` and `‹⇧`.

**Reply by evan-liu** [reply]
If you like them, putting them on a layer will be very useful. My layer is like 

```typescript
[
  withMapper({ 4: '⇥', 5: '⎋', 6: '⌘', 7: '⌥', 8: '⌃', 9: '⇧', 0: '⇪' })(
    (k, v) => map(k).toPaste(v),
  ),
  withMapper(['←', '→', '↑', '↓', '␣', '⏎', '⌫', '⌦'])((k) =>
    map(k).toPaste(k),
  ),
  { ',': toPaste('‹'), '.': toPaste('›') },
]
```
